src/main/java/net/snowflake/client/jdbc/ConnectionHandler.java

```diff
@@ -0,0 +1,42 @@
+package net.snowflake.client.jdbc;
+
+import java.sql.*;
+import java.util.Properties;
+import net.snowflake.client.core.BaseSFSession;
+import net.snowflake.client.core.SFBaseResultSet;
+
+/**
+ * Factory class that presents the implementation of a Snowflake Connection. This allows for
+ * alternate definitions of SFSession, SFStatement, and SFResultSet, (representing the 'physical'
+ * implementation layer) that can share high-level code.
+ */
+public interface ConnectionHandler {
+
+  /**
+   * Whether this Connection supports asynchronous queries. If yes, createAsyncResultSet may be
+   * called.
+   */
+  boolean supportsAsyncQuery();
+
+  /** Initializes the SnowflakeConnection */
+  void initializeConnection(String url, Properties info) throws SQLException;
+
+  /** Gets the SFSessionInterface implementation for this connection implementation */
+  BaseSFSession getSFSession();
+
+  /** Creates a result set from a query id. */
+  ResultSet createResultSet(String queryID, Connection connection) throws SQLException;
+
+  /**
+   * Creates a SnowflakeResultSet from a base SFBaseResultSet for this connection implementation.
+   */
+  SnowflakeBaseResultSet createResultSet(SFBaseResultSet resultSet, Statement statement)
```

Comment:
When would these createResultSet functions be used? I know we have one for an AsyncResultSet that can be created based on a query ID, but for a synchronous query, I don't recall a situation where we create a resultset without calling statement.executeQuery().

Reply:
See https://github.com/snowflakedb/snowflake-jdbc/pull/436/files#diff-3e78e7e0ce4b592a068d8e3eeb87fa3f58f8e4b313ff740ca0bc4b0d6e90ef8fR234 from the other closed PR that includes all three interfaces. It's called from within `Statement`. It's a slightly out-of-date PR, but this functionality was stripped from this PR so that I could split the interfaces into different pull requests.